fix(core): Fix PKCE OAuth2 flows to work without client secret

nealzhu3 opened the pull request:
## Summary

This PR fixes critical issues with PKCE (Proof Key for Code Exchange) OAuth2 implementation that prevented it from working correctly for public clients, as defined in [RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636).

### Problems Fixed:
**Token refresh fails for public PKCE clients**: The refresh token flow incorrectly requires a client secret, violating RFC 7636 which specifies that public clients using PKCE should not require client secrets

### Changes:
- Updated token refresh logic to handle public clients without requiring client secret
- Modified OAuth2 client creation to conditionally include client secret only when provided

### How to Test:
1. **Automated tests pass** - All unit tests for PKCE flows and OAuth2 credentials are included
2. **Manual testing**:
   - Create a new OAuth2 credential with PKCE grant type
   - Connect to an OAuth2 provider using PKCE
   - Verify that token refresh works correctly (with or without client secret)
 <img width="1236" height="845" alt="Screenshot 2025-09-16 at 11 41 39" src="https://github.com/user-attachments/assets/f835945b-7e9f-4220-87dc-7b7a9a5d0da0" />


## Related Linear tickets, Github issues, and Community forum posts

<!-- Add any related issues here if they exist -->
<!-- Example: Fixes #1234 -->

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md))
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included.
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

nealzhu3 committed:
feat(core): Add PKCE support for OAuth2 flows

Add support for PKCE (Proof Key for Code Exchange) OAuth2 flows to enable
secure authentication for public clients without requiring client secrets.

Changes:
- Make client secret optional in OAuth2 token refresh flows
- Update credential configuration to support PKCE grant type
- Add proper authentication handling for both confidential and public clients

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=19601) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=19601) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=19601) it.</sub>

Joffcom commented:
Hey @nealzhu3,

Thank you for your contribution. We appreciate the time and effort you’ve taken to submit this pull request.

Before we can proceed, please ensure the following:
	•	Tests are included for any new functionality, logic changes or bug fixes.
	•	The PR aligns with our contribution guidelines.

Regarding new nodes:
We no longer accept new nodes directly into the core codebase. Instead, we encourage contributors to follow our Community Node Submission Guide to publish nodes independently.

If your node integrates with an AI service that you own or represent, please email nodes@n8n.io and we will be happy to discuss the best approach.

About review timelines:
This PR has been added to our internal tracker as "GHC-4432". While we plan to review it, we are currently unable to provide an exact timeframe. Our goal is to begin reviews within a month, but this may change depending on team priorities. We will reach out when the review begins.

Thank you again for contributing to n8n.

nealzhu3 committed:
test(core): Add comprehensive tests for PKCE OAuth2 flows

- Add unit tests for PKCE authorization and token refresh flows
- Add integration tests for complete PKCE E2E flow
- Add credential schema tests for PKCE grant type configuration
- Test both public clients (no secret) and confidential clients
- Ensure proper handling of code_challenge and code_verifier parameters

nealzhu3 committed:
refactor(credentials): Remove client secret requirement for PKCE grant type

- Deleted client secret property from OAuth2Api credentials configuration
- Updated tests to reflect the removal of client secret as an optional field for PKCE
- Ensured that the changes align with the new PKCE flow implementation

nealzhu3 committed:
refactor(credentials): Remove unused ICredentialType import from OAuth2Api tests

nealzhu3 committed:
refactor(credentials): Simplify OAuth2 client creation and remove deprecated test file

- Refactored the OAuth2 client creation to directly use the ClientOAuth2 constructor with credentials.
- Removed the client secret condition for PKCE flows as it is no longer required.
- Deleted the outdated OAuth2Api credentials test file to clean up the codebase.

nealzhu3 committed:
fix lint

nealzhu3 committed:
fix lint

nealzhu3 committed:
refactor(tests): Update TypeScript ignore comments to expect-error for unused parameters

nealzhu3 commented:
> Hi @nealzhu3,
> 
> The typecheck and lint look good now, unfortunately the tests are now broken. Could you take a look at these again please?

Ah, the uri param is needed  because of nock's function signature. I removed it as it was causing the lint issue. I added back and `@ts-expect-error`. @afitzek please help me trigger the build again please.

afitzek reviewed (commented):
Hi @nealzhu3, 

Code looks good, just two small comments for the tests. 👍  

Please make sure to sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=19601).

nealzhu3 committed:
refactor(tests): Update unused parameter naming in PKCE flow tests

afitzek reviewed (approved):
Looks good to me. Thank you @nealzhu3 🙏
